src/time_integrator.f90

```diff
@@ -5,73 +5,270 @@ module m_time_integrator
 
   implicit none
 
+  private adams_bashforth, runge_kutta
+
   type :: time_intg_t
-    integer :: istep, nsteps, nsubsteps, order, nvars, nolds
+    integer :: method, istep, istage, order, nstep, nstage, nvars, nolds
+    real(dp) :: coeffs(4, 4)
+    real(dp) :: rk_b(4, 4)
+    real(dp) :: rk_a(3, 3, 4)
     type(flist_t), allocatable :: olds(:, :)
+    type(flist_t), allocatable :: curr(:)
+    type(flist_t), allocatable :: deriv(:)
     class(base_backend_t), pointer :: backend
     class(allocator_t), pointer :: allocator
   contains
     procedure :: step
+    procedure :: runge_kutta
+    procedure :: adams_bashforth
   end type time_intg_t
 
   interface time_intg_t
-    module procedure constructor
+    module procedure init
   end interface time_intg_t
 
 contains
 
-  function constructor(backend, allocator, nvars)
+  function init(backend, allocator, method, nvars)
     implicit none
 
-    type(time_intg_t) :: constructor
+    type(time_intg_t) :: init
     class(base_backend_t), pointer :: backend
     class(allocator_t), pointer :: allocator
+    integer, intent(in), optional :: method
     integer, intent(in), optional :: nvars
 
     integer :: i, j
 
-    constructor%backend => backend
-    constructor%allocator => allocator
+    ! initialize Runge-Kutta coefficients
+    ! rk1
+    init%rk_a(:, 1, 1) = [0.0_dp, 0._dp, 0._dp]
+    init%rk_a(:, 2, 1) = [0.0_dp, 0._dp, 0._dp]
+    init%rk_a(:, 3, 1) = [0.0_dp, 0._dp, 0._dp]
+    init%rk_b(:, 1) = [1._dp, 0._dp, 0._dp, 0._dp]
+
+    ! rk2
+    init%rk_a(:, 1, 2) = [0.5_dp, 0._dp, 0._dp]
+    init%rk_a(:, 2, 2) = [0.0_dp, 0._dp, 0._dp]
+    init%rk_a(:, 3, 2) = [0.0_dp, 0._dp, 0._dp]
+    init%rk_b(:, 2) = [0._dp, 1._dp, 0._dp, 0._dp]
+
+    ! rk3
+    init%rk_a(:, 1, 3) = [0.5_dp, 0._dp, 0._dp]
+    init%rk_a(:, 2, 3) = [0.0_dp, 3._dp/4._dp, 0._dp]
+    init%rk_a(:, 3, 3) = [0.0_dp, 0._dp, 0._dp]
+    init%rk_b(:, 3) = &
+      [2._dp/9.0_dp, 1._dp/3._dp, 4._dp/9._dp, 0._dp]
+
+    ! rk4
+    init%rk_a(:, 1, 4) = [0.5_dp, 0._dp, 0._dp]
+    init%rk_a(:, 2, 4) = [0._dp, 0.5_dp, 0._dp]
+    init%rk_a(:, 3, 4) = [0._dp, 0._dp, 1._dp]
+    init%rk_b(:, 4) = &
+      [1._dp/6._dp, 1._dp/3._dp, 1._dp/3._dp, 1._dp/6._dp]
```

Comment:
Currently, general Butcher coefficient storage is used for the explicit RK implementation. If we restrict the RK schemes from order 1 to 4, we can simplify this by only storing the diagonal of `rk_a`.

Reply:
Also, several alternatives exist for RK2, RK3, and RK4. We should agree on which version(s) of RK we want to support (SSP, Low storage, etc.) for each order.